### ISSUE TITLE
release ordering in the plugin site is a bit strange.

### ISSUE THREAD

**Issue by jtnord** [jenkins-infra/plugin-site#1193]
https://plugins.jenkins.io/kubernetes-credentials-provider/#releases lists release `0.18-1` at the top of the page - but that is older (both in date time, and maven/jenkins version comparison) than `0.20`

https://github.com/jenkinsci/kubernetes-credentials-provider-plugin/releases shows an expected order, so I guess this is something in the plugin-site code.

Expected behaviour

in a plugins releases releases are ordered by either :
* plugin version comparison (which is 99% the same as maven version ordering) (preferred)
* date order

Actual behaviour

* the plugin release version is not ordered by either of the expected ways

eg. for the [kuberneted crednetials provider plugin](https://plugins.jenkins.io/kubernetes-credentials-provider/#releases) I would expect the order to be `0.20` -> `0.18-1` -> `0.17`

<img width="692" alt="image" src="https://user-images.githubusercontent.com/494726/166819905-66253fb5-6a6f-4f03-8504-7008db834ba6.png">

**Comment by halkeye** [comment]
Huh, thats very weird.

https://github.com/jenkins-infra/plugin-site/blob/1031d57d35ddc738228485d6173a9b50a67d66dd/plugins/plugin-site/src/templates/plugin.jsx#L296 is where we do the graphql against the data we pulled from https://updates.jenkins.io/current/plugin-versions.json

its sorting by buildDate which should put 0.20 at the top.

```
$ curl -qs 'https://updates.jenkins.io/current/plugin-versions.json' | jq '.plugins["kubernetes-credentials-provider"]["0.20"].buildDate'
"Jun 23, 2021"
$ curl -qs 'https://updates.jenkins.io/current/plugin-versions.json' | jq '.plugins["kubernetes-credentials-provider"]["0.18-1"].buildDate'
"Mar 30, 2021"
```

so for triage purposes, yes this seems to be a bug, but no obvious reason as to why

**Comment by zbynek** [comment]
@halkeye your link to code that sorts by date is from 4th may, even though in december 2021 you changed it to sorting by padded version (https://github.com/jenkins-infra/plugin-site/commit/ee271fd10350a53738b16b20001c1f165bca39a0 ) to fix https://github.com/jenkins-infra/plugin-site/issues/914

Problem with the version padding is that it works well with semver, but breaks when versions use `-`, in this case
0.18-1 gets padded to `0_018-1`
0.20&nbsp;&nbsp;&nbsp;    gets padded to `0_00020`

The fix in this case is simple (split on both `.` and `-`), but covering all possible versioning schemes may be hard. Just wondering if we want to improve the version sorting algo or wait for https://github.com/jenkins-infra/update-center2/pull/636 to be merged and use timestamp for sorting. CC @daniel-beck

**Comment by daniel-beck** [comment]
> Just wondering if we want to improve the version sorting algo or wait for https://github.com/jenkins-infra/update-center2/pull/636 to be merged and use timestamp for sorting

There's already a date in a kinda crappy format in the JSON, just doesn't have any better resolution than date (no time). Should still help with 11 months vs. 1 year.

**Comment by zbynek** [comment]
@daniel-beck yes, that's what was used before #914 was fixed (that issue is about 2 releases of CD enabled plugin that happened within 24 hours). 

The oly difference between version sort and timestamp sort is when the plugin has 2 release lines and 1.X gets released after 2.Y. I'm wondering what's the betterr behaviour there from user's perspective (IIRC GitHub sorts releases by date too).

**Comment by zbynek** [comment]
OK, so @jglick @jtnord @daniel-beck @MarkEWaite @halkeye what do we want: releases sorted by version or by timestamp? 

There are ~400 pairs of consecutive releases where the chronologically newer one has lower version (using a slightly improved but not perfect JS implementation of maven version comparison): https://gist.github.com/zbynek/46b7c3a6afdf4f912c7ba1d3648780f6

**Comment by halkeye** [comment]
I haven't found too many plugins that do multiple release lines. Blueocean in the old days did and it really was confusing.

I would like to stick with however update center does it, which I guess is by version number?

**Comment by MarkEWaite** [comment]
> I would like to stick with however update center does it

That is my preference as well.  The current ordering on the [git plugin page](https://plugins.jenkins.io/git/#releases) shows 4.10.0 released 10 months ago and then 4.9.4 released 2 weeks ago.  I like that ordering.  Release 4.10.0 is "greater than" release 4.9.4 for me, so it makes sense that 4.10.0 is higher on the page than 4.9.4 even though 4.9.4 was released after 4.10.0

![Screenshot 2022-08-25 121020](https://user-images.githubusercontent.com/156685/186738423-bdff7fc6-f82b-4afe-8864-81e83e3e7f32.png)
